text/0000-union.md

```diff
@@ -0,0 +1,371 @@
+- Feature Name: `union`
+- Start Date: 2015-12-29
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Provide native support for C-compatible unions, defined via a new "contextual
+keyword" `union`, without breaking any existing code that uses `union` as an
+identifier.
+
+# Motivation
+[motivation]: #motivation
+
+Many FFI interfaces include unions.  Rust does not currently have any native
+representation for unions, so users of these FFI interfaces must define
+multiple structs and transmute between them via `std::mem::transmute`.  The
+resulting FFI code must carefully understand platform-specific size and
+alignment requirements for structure fields.  Such code has little in common
+with how a C client would invoke the same interfaces.
+
+Introducing native syntax for unions makes many FFI interfaces much simpler and
+less error-prone to write, simplifying the creation of bindings to native
+libraries, and enriching the Rust/Cargo ecosystem.
+
+A native union mechanism would also simplify Rust implementations of
+space-efficient or cache-efficient structures relying on value representation,
+such as machine-word-sized unions using the least-significant bits of aligned
+pointers to distinguish cases.
+
+The syntax proposed here recognizes `union` as though it were a keyword when
+used to introduce a union declaration, *without* breaking any existing code
+that uses `union` as an identifier.  Experiments by Niko Matsakis demonstrate
+that recognizing `union` in this manner works unambiguously with zero conflicts
+in the Rust grammar.
+
+To preserve memory safety, accesses to union fields may only occur in unsafe
+code.  Commonly, code using unions will provide safe wrappers around unsafe
+union field accesses.
+
+# Detailed design
+[design]: #detailed-design
+
+## Declaring a union type
+
+A union declaration uses the same field declaration syntax as a struct
+declaration, except with `union` in place of `struct`.
+
+```rust
+union MyUnion {
+    f1: u32,
+    f2: f32,
+}
+```
+
+`union` implies `#[repr(C)]` as the default representation.
```

Comment:
Something we should maybe consider is removing unused fields/variants aggressively based on privacy.
It's only tangential, but I've realized that in a state machine, you can get away with state bloat if you get to optimize the state ADT representation based on remaining uses in optimized MIR.